xbmc/cores/AudioEngine/Sinks/pipewire/AESinkPipewire.cpp

```diff
@@ -607,7 +607,8 @@ unsigned int CAESinkPipewire::AddPackets(uint8_t** data, unsigned int frames, un
 
   } while (true);
 
-  m_stream->TriggerProcess();
+  if (m_stream->IsDriving())
```

Comment:
Due to the design - the two commands are "non atomic". Means you could be scheduled away and condition when you call TriggerProcess might be the other way round already and your log comes back. As those two belong together, they need to be protected with a critical section.

Key question is: Is it worth it?

Reply:
Also from the PW docu:
Streams that are not drivers according to [pw_stream_is_driving()](https://docs.pipewire.org/group__pw__stream.html#ga3ddd332c45aa95250d3d1d26cf01452a) can also call this method. The result is that a RequestProcess event is sent to the driver. If the graph is lazy scheduling according to [pw_stream_is_lazy()](https://docs.pipewire.org/group__pw__stream.html#gac987fb3ab601efd721dbd912e21b9136), this might result in a graph cycle by the driver. If the graph is not lazy scheduling and the stream is not a driver, this method will have no effect.

We are very well allowed to call it - even if we are not driving ...

Reply:
That is an interesting find. When keep audio alive packages are queued only, they would have no effect.

One reason why Kodi wants to be a driver is the delay management. We want to know and feel at all times that audio was consumed under the hood. Also by design audio engine in Kodi has a driving loop itself. The Kodi audio sink thread is only blocked by the PW sink implementation. So it is clear why driver design was chosen.

Need to check again where the bug is. The stream noise case rings my alarm bells in a sense of that stream trigger is anyways not fully working as we thought it works.

Thanks for looking into this!

Reply:
It might be that PW does an idle detection and unloads us as a driver when AE stops adding samples when being in its idle state without being alive. That way you should only get it when pressing a button.

It might be that we have to reset the driving state. E.g. when it removed our driver state, we have to ask it again?  If (!driving) make it drive() followed by process?

Reply:
There is no direct possibility (e.g. set_driving) available. There is the stream properties that could update it - but honestly - I think something is wrong here. Do you see that PW reports idle and unloads the sink?

Reply:
Okay. That basically means, we properly have to issue timeouts in pipewire itself when the sink is not being used.

For the menu sounds: Yes, that is the reason we implemented keep audio alive in first place as HDMI receivers and others would simply shutdown.